### ISSUE TITLE
TrimKey not called on struct of Map

### ISSUE THREAD

**Issue by enrico-giovanni** [getnamo/socketio-client-ue4#59]
ue4 4.16.3
sio 0.6.2
environment: Blueprint

This JSON is what I receive on the server:
```json
{
  "actions": {
    "my_key": {
      "value_5_0C62405F4F8CD0DE2EFE1C84058E954D": 123,
      "vector_8_CBC3A5A248B0C7EFE10514AF9D644328": {
        "x": 1,
        "y": 1,
        "z": 1
      }
    }
  }
}
```

when I Emit this Actions struct:
struct Actions, field 'actions' which is a map of String -> Action
struct Action, fields 'value' = Float, 'vector' = Vector

The two fields value and vector seem not properly trimmed

The BP nodes I'm using are:
```
variable get Actions -> Struct to JSON Object -> ToJSONValue -> Emit
```

**Comment by getnamo** [comment]
Interesting, can you provide an image of your blueprint node for setting the data?

**Comment by enrico-giovanni** [comment]
Here I call the server just to show the JSON:
![bp](https://user-images.githubusercontent.com/14308165/29216418-eef274a2-7eae-11e7-86e2-68393676b0c5.png)

This is the variable my_struct_with_map's definition:
![struct_with_map](https://user-images.githubusercontent.com/14308165/29216433-fbc9e64c-7eae-11e7-8b3c-49e1e54d4bea.png)

This is the type of the value-side of the map:
![simple_struct](https://user-images.githubusercontent.com/14308165/29216440-feaffce8-7eae-11e7-8e41-20bdbd0507ce.png)

This is the some values assigned to the my_struct_with_map variable:
![variable](https://user-images.githubusercontent.com/14308165/29216444-0222f40c-7eaf-11e7-9b8a-1063cd1a1c4d.png)

This is what the server receives:
```javascript
{
  "my_map": {
    "key_1": {
      "number_5_525DA0A94AA47E09B7EE258D9F1224B9": 1,
      "text_7_D6E1B42A479104B9D09BCF87B93E186D": "one"
    },
    "key_2": {
      "number_5_525DA0A94AA47E09B7EE258D9F1224B9": 2,
      "text_7_D6E1B42A479104B9D09BCF87B93E186D": "two"
    }
  }
}
```

This is the expected:
```javascript
{
  "my_map": {
    "key_1": {
      "number": 1,
      "text": "one"
    },
    "key_2": {
      "number": 2,
      "text": "two"
    }
  }
}
```

In your code I've seen two or three trimming functions, which I believe you use to trim the default selector name assigned by UE4.
Maybe you need to recurse into the struct and apply one of these functions.

Btw, congrats for this nice library :)

EJ

**Comment by getnamo** [comment]
Excellent this should be all I need. I'll make a pass on this for the next update. Those names btw only appear when you create bp structs (vs c++ ue4 structs), I assume this is Unreal's way to ensure the keys are unique and you are correct it appears my trimming function isn't looping through them correctly for nested maps.

**Comment by getnamo** [comment]
Pushed a fix for this in: https://github.com/getnamo/socketio-client-ue4/commit/f640f639a019f73f6068c16aba23104210ecb8fd

After initially checking the source, I realized I had never tested nor supported TMaps in BP Structs. Getting those to trim for UE4 struct with TMap -> JSON string was pretty simple  (a few lines around https://github.com/getnamo/socketio-client-ue4/commit/f640f639a019f73f6068c16aba23104210ecb8fd#diff-be1bcd60c471b6d04fc0dfb03a660fceR274), but to get the other way working (json string -> fill struct with tmaps) was much more complicated.

I settled for a way to mark tmap short names in my trimming mapping function and replacing them when the json value gets filled in later (https://github.com/getnamo/socketio-client-ue4/commit/f640f639a019f73f6068c16aba23104210ecb8fd#diff-be1bcd60c471b6d04fc0dfb03a660fceR429). Because UE4 can only contain TMaps of the same type this system works pretty darn well.

To test the system I made a very complicated nested struct with different types of TMaps (int/string), nested structs of nested struct, arrays etc... definitely overboard.

Top Level
![toplevelstruct](https://user-images.githubusercontent.com/542365/32423226-079a79dc-c29d-11e7-9c32-c13595b377b2.PNG)

Simple Struct
![simplestruct](https://user-images.githubusercontent.com/542365/32423218-f9db772e-c29c-11e7-95eb-7b5469514df5.PNG)

Nested Test Struct
![nestedstruct](https://user-images.githubusercontent.com/542365/32423236-19502852-c29d-11e7-97e5-d6bcc9358f74.PNG)

It's sub struct
![subnestedstruct](https://user-images.githubusercontent.com/542365/32423239-1eb5cb62-c29d-11e7-8d28-5110d1a9ec1f.PNG)

with a few default values tested

![ue4struct](https://user-images.githubusercontent.com/542365/32423043-1ab9b8c6-c29c-11e7-88ee-9c3a0b05ac20.PNG)

which went through a round trip of encoding and decoding

![testingsetup](https://user-images.githubusercontent.com/542365/32423035-098fbb4a-c29c-11e7-82d0-90be79ae1914.PNG)
 
it spits out valid json
```JSON
{  
   "integermap":{  
      "1.0":{  
         "vecKey":{  
            "x":1,
            "y":2,
            "z":3
         },
         "stringKey":"",
         "structKey":{  
            "testVar_3_asdlkajsd":false,
            "memberVar_5":[  
               4,
               5
            ],
            "stringKey":"I shouldn't clash!"
         },
         "arrayKey":[  

         ],
         "boolKey":true
      },
      "3.0":{  
         "vecKey":{  
            "x":0,
            "y":0,
            "z":0
         },
         "stringKey":"yolo",
         "structKey":{  
            "testVar_3_asdlkajsd":true,
            "memberVar_5":[  
               1.2000000476837158
            ],
            "stringKey":"I shouldn't clash!"
         },
         "arrayKey":[  
            {  
               "boolMeBro":[  
                  true,
                  false
               ]
            }
         ],
         "boolKey":false
      }
   },
   "stringfloatmap":{  
      "a string":-1,
      "another string":2.5339999198913574
   },
   "my_map":{  
      "key_1":{  
         "text":"one",
         "number":1
      },
      "key_2":{  
         "text":"two",
         "number":2
      }
   }
}
```
and the receiving struct correctly has the internal bp names set

![debugview](https://user-images.githubusercontent.com/542365/32423082-504f76f6-c29c-11e7-93d0-df0810ea69f3.PNG)

This should be now be a dependable path, but let me know if you find a use case that doesn't pass.

Will coalesce a few more fixes before pushing a binary 4.18 release.

**Comment by getnamo** [comment]
Closing issue. Re-open if any bugs appear.
